File: Day #1411/Solution.java
// Day 1411: Check if tree t is a subtree of tree s.
// Approach: recursive DFS — for each node of s try exact-match with t. O(|s|*|t|) time, O(h) space.
public class Solution {
    static class Node {
        int val;
        Node left, right;
        Node(int v) { val = v; }
    }

    static boolean sameTree(Node a, Node b) {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;
        return a.val == b.val && sameTree(a.left, b.left) && sameTree(a.right, b.right);
    }

    static boolean isSubtree(Node s, Node t) {
        if (s == null) return false;
        if (sameTree(s, t)) return true;
        return isSubtree(s.left, t) || isSubtree(s.right, t);
    }

    public static void main(String[] args) {
        Node s = new Node(3);
        s.left = new Node(4);
        s.right = new Node(5);
        s.left.left = new Node(1);
        s.left.right = new Node(2);

        Node t = new Node(4);
        t.left = new Node(1);
        t.right = new Node(2);

        System.out.println(isSubtree(s, t) ? "true" : "false");
    }
}

File: Day #1411/solution.rs
// Day 1411: Check if tree t is a subtree of tree s.
// Approach: recursive DFS — for each node of s try exact-match with t. O(|s|*|t|) time, O(h) space.
use std::rc::Rc;

struct Node {
    val: i32,
    left: Option<Rc<Node>>,
    right: Option<Rc<Node>>,
}

fn node(val: i32, left: Option<Rc<Node>>, right: Option<Rc<Node>>) -> Rc<Node> {
    Rc::new(Node { val, left, right })
}

fn same_tree(a: &Option<Rc<Node>>, b: &Option<Rc<Node>>) -> bool {
    match (a, b) {
        (None, None) => true,
        (Some(x), Some(y)) => {
            x.val == y.val && same_tree(&x.left, &y.left) && same_tree(&x.right, &y.right)
        }
        _ => false,
    }
}

fn is_subtree(s: &Option<Rc<Node>>, t: &Option<Rc<Node>>) -> bool {
    match s {
        None => false,
        Some(n) => same_tree(s, t) || is_subtree(&n.left, t) || is_subtree(&n.right, t),
    }
}

fn main() {
    let s = Some(node(
        3,
        Some(node(4, Some(node(1, None, None)), Some(node(2, None, None)))),
        Some(node(5, None, None)),
    ));
    let t = Some(node(4, Some(node(1, None, None)), Some(node(2, None, None))));
    println!("{}", if is_subtree(&s, &t) { "true" } else { "false" });
}
